### PR TITLE
Choose machine deployment by name or selector

### LINKED ISSUE #77: Allow selection of MachineDeployments when upgrading worker nodes
When using the `machine-deployment` scope, it would be nice if I was able to select a single instance and/or group of MachineDeployment(s) to upgrade. As the tool is written now, it will upgrade all MachineDeployments for a cluster in a single run. I would like to be able to specify the name of a MachineDeployment and/or provide a custom label selector of MachineDeployments to upgrade.

The use case is to have more control over the upgrade process. For example, I may want to avoid upgrading nodes with a taint of "finance" until I'm comfortable upgrading the rest the cluster. Or, I'd like to upgrade one MachineDeployment at a time to control the amount of churn in the cluster.

### PR DESCRIPTION
**Pull request by ncdc**
Add the ability to specify which machine deployemnt(s) to upgrade by
name or by label selector.

Fixes #77

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -78,6 +78,20 @@ func main() {
...
+		"",
+		"Name of a single machine deployment to upgrade",
+	)
+
+	root.Flags().StringVar(
+		&upgradeConfig.MachineDeployment.LabelSelector,
+		"machine-deployment-selector",
+		"",
+		"Label selector used to find machine deployments to upgrade",
+	)
```
Comment (detiber) [inline main.go]: Is it possible in cobra to mark these two flags as mutually exclusive?

File: pkg/upgrade/machine_deployment.go [hunk 2]
```diff
@@ -44,11 +77,15 @@ func (u *MachineDeploymentUpgrader) listMachineDeployments() (*clusterv1.Machine
 	u.log.Info("Listing machine deployments")
 
 	listOptions := []ctrlclient.ListOption{
-		ctrlclient.MatchingLabels{
-			clusterv1.MachineClusterLabelName: u.clusterName,
-		},
 		ctrlclient.InNamespace(u.clusterNamespace),
 	}
+
+	if u.selector == nil {
```
[the remark this answers is not in the document]
Comment (detiber) [reply pkg/upgrade/machine_deployment.go]: I'm torn, part of me wants to believe that people will choose the right thing, but if we can provide safety then we probably should.

[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/upgrade/machine_deployment.go]: Sure, I'll see what I can do

[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/upgrade/machine_deployment.go]: @detiber updated, ptal

File: pkg/upgrade/machine_deployment.go [hunk 3]
```diff
@@ -22,15 +26,55 @@ func NewMachineDeploymentUpgrader(log logr.Logger, config Config) (*MachineDeplo
...
+			Name:      u.name,
+		}
+
+		var machineDeployment clusterv1.MachineDeployment
+
+		if err := u.managerClusterClient.Get(context.TODO(), key, &machineDeployment); err != nil {
+			return errors.Wrapf(err, "error getting machine deployment %q", u.name)
+		}
+
+		machineDeployments = &clusterv1.MachineDeploymentList{Items: []clusterv1.MachineDeployment{machineDeployment}}
```
Comment (detiber) [inline pkg/upgrade/machine_deployment.go]: Should this validate that the returned machineDeployment belongs to u.ClusterName?

Comment (ncdc) [reply pkg/upgrade/machine_deployment.go]: Will do

**Comment by ncdc** [comment]
@detiber updated

**Comment by ncdc** [comment]
@detiber you want anyone else to review or ok to merge now?

**Pull request merged**